File: app/src/main/java/tw/com/chainsea/chat/view/chatroom/adapter/AdvisoryRoomAdapter.java
//package tw.com.chainsea.chat.view.chatroom.adapter;
//
//import android.content.Context;
//import android.view.LayoutInflater;
//import android.view.View;
//import android.view.ViewGroup;
//import android.widget.ImageView;
//import android.widget.TextView;
//
//import androidx.annotation.NonNull;
//import androidx.recyclerview.widget.RecyclerView;
//
//import com.google.common.base.Strings;
//import com.google.common.collect.Iterables;
//import com.google.common.collect.Lists;
//
//import java.util.Collections;
//import java.util.List;
//
//import tw.com.chainsea.android.common.json.JsonHelper;
//import tw.com.chainsea.android.common.text.StringHelper;
//import tw.com.chainsea.ce.sdk.bean.Entity;
//import tw.com.chainsea.ce.sdk.bean.PicSize;
//import tw.com.chainsea.ce.sdk.bean.ProcessStatus;
//import tw.com.chainsea.ce.sdk.bean.room.ChatRoomEntity;
//import tw.com.chainsea.ce.sdk.bean.room.ChatRoomType;
//import tw.com.chainsea.ce.sdk.bean.todo.TodoEntity;
//import tw.com.chainsea.ce.sdk.bean.todo.TodoStatus;
//import tw.com.chainsea.ce.sdk.database.sp.TokenPref;
//import tw.com.chainsea.ce.sdk.database.sp.UserPref;
//import tw.com.chainsea.ce.sdk.event.EventMsg;
//import tw.com.chainsea.ce.sdk.event.MsgConstant;
//import tw.com.chainsea.ce.sdk.reference.UserProfileReference;
//import tw.com.chainsea.ce.sdk.service.AvatarService;
//import tw.com.chainsea.chat.R;
//import tw.com.chainsea.chat.databinding.ItemAdvisorySmallRoomBinding;
//import tw.com.chainsea.chat.style.RoomThemeStyle;
//import tw.com.chainsea.chat.util.Unit;
//import tw.com.chainsea.chat.util.UnreadUtil;
//import tw.com.chainsea.custom.view.image.CircleImageView;
//import tw.com.chainsea.custom.view.recyclerview.itemdecoration.ItemBaseViewHolder;
//import tw.com.chainsea.custom.view.recyclerview.itemdecoration.ItemNoSwipeViewHolder;
//
///**
// * Create by evan on 1/15/21
// *
// * @author Evan Wang
// * date 1/15/21
// */
//public class AdvisoryRoomAdapter extends RecyclerView.Adapter<ItemBaseViewHolder<Entity>> {
//    private final List<Entity> entities = Lists.newArrayList();
//
//    private final List<ChatRoomEntity> roomEntities = Lists.newArrayList();
//    private final List<ChatRoomEntity> memberRoomEntities = Lists.newArrayList();
//    private final List<TodoEntity> todoEntities = Lists.newArrayList();
//    private Context context;
//    private String selfId = "";
//    private String roomId = "";
//
//    RoomThemeStyle themeStyle = RoomThemeStyle.UNDEF;
//
//    private OnAdvisoryListener onAdvisoryListener;
//
//    long now = System.currentTimeMillis();
//
//    @NonNull
//    @Override
//    public ItemBaseViewHolder<Entity> onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
//        if (this.context == null) {
//            this.context = parent.getContext();
//            this.selfId = TokenPref.getInstance(this.context).getUserId();
//        }
//        ItemAdvisorySmallRoomBinding binding = ItemAdvisorySmallRoomBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
//        if (viewType == 0) {
//            return new LeftViewHolder(binding);
//        } else {
//            return new SmallRoomViewHolder(binding);
//        }
//    }
//
//    @Override
//    public int getItemViewType(int position) {
//        Entity entity = this.entities.get(position);
//        if ("ChatRoomEntity".equals(entity.getClassName())) {
//            return 1;
//        } else {
//            return 0;
//        }
//    }
//
//    @Override
//    public void onBindViewHolder(@NonNull ItemBaseViewHolder<Entity> holder, int position) {
//        Entity entity = this.entities.get(position);
//        holder.onBind(entity, 0, position);
//    }
//
//    @Override
//    public int getItemCount() {
//        return this.entities.size();
//    }
//
//    private Context getContext() {
//        return this.context;
//    }
//
//    private String getUserId() {
//        return this.selfId;
//    }
//
//    public AdvisoryRoomAdapter setRoomThemeStyle(RoomThemeStyle themeStyle) {
//        this.themeStyle = themeStyle;
//        return this;
//    }
//
//    public AdvisoryRoomAdapter setRoomId(String roomId) {
//        this.roomId = StringHelper.getString(roomId, "").toString();
//        return this;
//    }
//
//    public AdvisoryRoomAdapter setServiceMemberRoom(ChatRoomEntity entity) {
//        if (entity != null) {
//            //this.memberRoomEntities.clear();
//            this.memberRoomEntities.add(entity);
//        }
//        return this;
//    }
//
//    public void clearRoomData() {
//        this.roomEntities.clear();
//        notifyDataSetChanged();
//    }
//    public AdvisoryRoomAdapter removeRoomData(String roomId) {
//        this.roomEntities.remove(ChatRoomEntity.Build().id(roomId).build());
//        return this;
//    }
//
//    public AdvisoryRoomAdapter setRoomData(List<ChatRoomEntity> list) {
////        for (ChatRoomEntity entity : list) {
////            this.roomEntities.remove(entity);
//            this.roomEntities.addAll(list);
////        }
//        return this;
//    }
//
//    public AdvisoryRoomAdapter setTodoData(List<TodoEntity> list) {
//        for (TodoEntity entity : list) {
//            this.todoEntities.remove(entity);
//            this.todoEntities.add(entity);
//        }
//        return this;
//    }
//
//    public AdvisoryRoomAdapter setListener(OnAdvisoryListener onAdvisoryListener) {
//        this.onAdvisoryListener = onAdvisoryListener;
//        return this;
//    }
//
//    public void filter() {
//
//    }
//
//    public void sort() {
//        if (roomEntities != null){
//        Collections.sort(this.roomEntities);
//        }
//    }
//
//    private void assembly() {
//        this.entities.clear();
//        if (!this.memberRoomEntities.isEmpty()) {
//            ChatRoomEntity original = Iterables.getFirst(this.memberRoomEntities, null);
//            this.entities.add(original);
//        }
//
//        if (!this.todoEntities.isEmpty()) {
//            this.now = System.currentTimeMillis();
//            boolean isShow = false;
//            for (TodoEntity t : this.todoEntities) {
//                if (t.getRemindTime() > 0 && t.getRemindTime() < this.now && TodoStatus.PROGRESS.equals(t.getStatus())) {
//                    isShow = true;
//                    break;
//                }
//            }
//
//            if (isShow) {
//                TodoEntity original = Iterables.getFirst(this.todoEntities, null);
//                this.entities.add(original);
//            }
//        }
//
//        if (!this.roomEntities.isEmpty()) {
//            this.entities.addAll(this.roomEntities);
//        }
//    }
//
//    public void refreshData() {
//        this.now = System.currentTimeMillis();
//        filter();
//        sort();
//        assembly();
//        if (this.onAdvisoryListener != null) {
//            this.onAdvisoryListener.onItemCount(this.entities.size());
//        }
//        notifyDataSetChanged();
//    }
//
//    class LeftViewHolder extends ItemNoSwipeViewHolder<Entity> {
//
//        private ItemAdvisorySmallRoomBinding binding;
//
//        public LeftViewHolder(ItemAdvisorySmallRoomBinding binding) {
//            super(binding.getRoot());
//            this.binding = binding;
//            binding.civIcon.setVisibility(View.VISIBLE);
//            binding.ivTodoIcon.setVisibility(View.GONE);
//        }
//
//        @Override
//        public void onBind(Entity entity, int section, int position) {
//            if (entity instanceof TodoEntity) {
//                binding.civIcon.setImageResource(themeStyle.getTodoIconResId());
////                civIcon.setImageResource(R.drawable.res_check_list_circle_def);
////                civIcon.setColorFilter(ContextCompat.getColor(context, themeStyle.getMainColor()));
//                TodoEntity t = (TodoEntity) entity;
//                int count = 0;
//                now = System.currentTimeMillis();
//                for (TodoEntity todo : todoEntities) {
//                    if (todo.getRemindTime() > 0 && todo.getRemindTime() < now && TodoStatus.PROGRESS.equals(todo.getStatus())) {
//                        count++;
//                    }
//                }
//                if (count > 0) {
//                    binding.tvUnread.setVisibility(View.VISIBLE);
//                    binding.tvUnread.setText(String.valueOf(count));
//                } else {
//                    binding.tvUnread.setVisibility(View.GONE);
//                }
//
//                itemView.setOnClickListener(v -> {
//                    if (onAdvisoryListener != null) {
//                        onAdvisoryListener.onTodoSettingClick(todoEntities);
//                    }
//                });
//
//            }
//        }
//    }
//
//    class SmallRoomViewHolder extends ItemNoSwipeViewHolder<Entity> {
//        private ItemAdvisorySmallRoomBinding binding;
//
//        public SmallRoomViewHolder(ItemAdvisorySmallRoomBinding binding) {
//            super(binding.getRoot());
//            this.binding = binding;
//            binding.ivTodoIcon.setVisibility(View.GONE);
//        }
//
//        @Override
//        public void onBind(Entity entity, int section, int position) {
//            if (entity instanceof ChatRoomEntity) {
//                ChatRoomEntity chatRoomEntity = (ChatRoomEntity) entity;
//                if (chatRoomEntity.getUnReadNum() <= 0) {
//                    binding.tvUnread.setVisibility(View.GONE);
//                } else {
//                    binding.tvUnread.setVisibility(View.VISIBLE);
//                }
//
//                if (ChatRoomType.serviceMember.equals(chatRoomEntity.getType())) {
//                    binding.civIcon.setImageResource(themeStyle.getServiceMemberIconResId());
//
//                    itemView.setOnClickListener(v -> {
//                        //服務成員聊天室點擊事件
//                        if (onAdvisoryListener != null) {
//                            onAdvisoryListener.onServiceMemberRoomClick(chatRoomEntity);
//                        }
//                    });
//                } else {
//                    if (ChatRoomType.discuss.equals(chatRoomEntity.getType()) || (ChatRoomType.group.equals(chatRoomEntity.getType()) && Strings.isNullOrEmpty(chatRoomEntity.getAvatarId()))) {
//                        if (chatRoomEntity.getMemberAvatarData() == null || chatRoomEntity.getMemberAvatarData().isEmpty()) {
//                            chatRoomEntity.setMemberAvatarData(UserProfileReference.getMemberAvatarData(null, chatRoomEntity.getId(), getUserId(), 4));
//                        }
//                    }
//
//                    AvatarService.post(getContext(), chatRoomEntity.getServiceNumberAvatarId(), PicSize.SMALL, binding.civIcon, R.drawable.default_avatar);
//
//                    itemView.setOnClickListener(v -> {
//                        if (onAdvisoryListener != null) {
//                            onAdvisoryListener.onPrivateRoomClick(chatRoomEntity);
//                        }
//                    });
//                }
//            }
//        }
//    }
//
//    public interface OnAdvisoryListener {
//
//        void onServiceMemberRoomClick(ChatRoomEntity roomEntity);
//
//        void onTodoSettingClick(List<TodoEntity> entities);
//
//        void onPrivateRoomClick(ChatRoomEntity roomEntity);
//
//        void onItemCount(int count);
//    }
//
//    public void handleEvent(EventMsg eventMsg) {
//        switch (eventMsg.getCode()) {
//            case MsgConstant.SESSION_UPDATE_FILTER:
//            case MsgConstant.REFRESH_ROOM_BY_LOCAL:
//                ChatRoomEntity refreshEntity = JsonHelper.getInstance().from(eventMsg.getString(), ChatRoomEntity.class);
//                int refreshIndex = roomEntities.indexOf(refreshEntity);
//                if (refreshIndex != -1) {
//                    roomEntities.remove(refreshEntity);
//                    roomEntities.add(refreshIndex, refreshEntity);
//                }
//                refreshIndex = memberRoomEntities.indexOf(refreshEntity);
//                if (refreshIndex != -1) {
//                    memberRoomEntities.remove(refreshEntity);
//                    memberRoomEntities.add(refreshIndex, refreshEntity);
//                }
//                refreshData();
//                break;
//            case MsgConstant.UPDATE_TODO_EXPIRED_COUNT_EVENT:
//                refreshData();
//                break;
//            case MsgConstant.UI_NOTICE_TODO_REFRESH:
//                TodoEntity entity = JsonHelper.getInstance().from(eventMsg.getString(), TodoEntity.class);
//                if (TodoStatus.DELETED.equals(entity.getStatus())) {
//                    todoEntities.remove(entity);
//                    refreshData();
//                } else if (!ProcessStatus.UN_SYNC_DELETE.equals(entity.getProcessStatus())) {
//                    String personRoomId = UserPref.getInstance(context).getPersonRoomId();
//                    if (this.roomId.equals(personRoomId) && Strings.isNullOrEmpty(entity.getRoomId())) {
//                        todoEntities.remove(entity);
//                        todoEntities.add(entity);
//                    }
//
//                    if (this.roomId.equals(entity.getRoomId())) {
//                        todoEntities.remove(entity);
//                        todoEntities.add(entity);
//                    }
//                    refreshData();
//                }
//                break;
//
//        }
//    }
//
//
//}

File: app/src/main/java/tw/com/chainsea/chat/view/chatroom/adapter/AdvisoryRoomAdapter.kt
package tw.com.chainsea.chat.view.chatroom.adapter

import android.annotation.SuppressLint
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.activity.result.ActivityResultLauncher
import androidx.core.os.bundleOf
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import tw.com.chainsea.ce.sdk.bean.room.ChatRoomEntity
import tw.com.chainsea.ce.sdk.bean.room.ChatRoomType
import tw.com.chainsea.ce.sdk.bean.todo.TodoEntity
import tw.com.chainsea.ce.sdk.reference.ChatRoomReference
import tw.com.chainsea.chat.R
import tw.com.chainsea.chat.chatroomfilter.ChatRoomFilterActivity
import tw.com.chainsea.chat.config.BundleKey
import tw.com.chainsea.chat.databinding.ItemAdvisorySmallRoomBinding
import tw.com.chainsea.chat.databinding.ItemAdvisoryTodoBinding
import tw.com.chainsea.chat.ui.activity.ChatActivity
import tw.com.chainsea.chat.util.IntentUtil
import tw.com.chainsea.chat.util.ThemeHelper
import tw.com.chainsea.chat.view.consultai.ConsultAIActivity

class AdvisoryRoomAdapter(
    val consultAILauncher: ActivityResultLauncher<Intent>
) : ListAdapter<SmallRoomData, RecyclerView.ViewHolder>(AdvisoryRoomAdapterDiffCallBack()) {
    companion object {
        const val SERVICE_CONSULT_ROOM = 1
        const val TODO = 2
    }

    fun setData(data: List<SmallRoomData>) {
        submitList(data)
    }

    fun shouldNotifyUnreadIcon(roomId: String) =
        CoroutineScope(Dispatchers.IO).launch {
            currentList.forEachIndexed { index, smallRoomData ->
                if (smallRoomData.roomId == roomId) {
                    smallRoomData.unReadNum = 1
                    withContext(Dispatchers.Main) {
                        notifyItemChanged(index)
                    }
                    return@launch
                }
            }
        }

    override fun getItemViewType(position: Int): Int =
        if (getItem(position).todoList.isEmpty()) {
            SERVICE_CONSULT_ROOM
        } else {
            TODO
        }

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): RecyclerView.ViewHolder {
        val holder =
            when (viewType) {
                SERVICE_CONSULT_ROOM -> {
                    val binding = ItemAdvisorySmallRoomBinding.inflate(LayoutInflater.from(parent.context), parent, false)
                    SmallRoomViewHolder(binding)
                }

                else -> {
                    val binding = ItemAdvisoryTodoBinding.inflate(LayoutInflater.from(parent.context), parent, false)
                    TodoSmallIconViewHolder(binding)
                }
            }
        return holder
    }

    override fun onBindViewHolder(
        holder: RecyclerView.ViewHolder,
        position: Int
    ) {
        when (getItemViewType(position)) {
            SERVICE_CONSULT_ROOM -> {
                (holder as SmallRoomViewHolder).bind(getItem(position))
            }

            else -> {
                (holder as TodoSmallIconViewHolder).bind(getItem(position))
            }
        }
    }

    inner class SmallRoomViewHolder(
        val binding: ItemAdvisorySmallRoomBinding
    ) : RecyclerView.ViewHolder(binding.root) {
        fun bind(data: SmallRoomData) {
            binding.tvUnread.visibility = View.GONE
            when (data.roomType) {
                ChatRoomType.serviceMember -> {
                    binding.tvUnread.visibility = if (data.unReadNum > 0) View.VISIBLE else View.GONE
                    binding.civIcon.setImageResource(data.avatarId.toInt())
                }

                ChatRoomType.services -> {
                    var serviceNumberChatRoom: ChatRoomEntity? = null
                    CoroutineScope(Dispatchers.IO).launch {
                        serviceNumberChatRoom = ChatRoomReference.getInstance().findById(data.roomId)
                        serviceNumberChatRoom?.let {
                            withContext(Dispatchers.Main) {
                                binding.tvUnread.visibility = if (it.unReadNum > 0) View.VISIBLE else View.GONE
                                binding.civIcon.loadAvatarIcon(data.avatarId, it.name, data.serviceNumberId)
                            }
                        } ?: run {
                            withContext(Dispatchers.Main) {
                                binding.civIcon.loadAvatarIcon(data.avatarId, "", data.serviceNumberId)
                            }
                        }
                    }
                }

                else -> {
                    binding.civIcon.setImageResource(R.drawable.ic_ai_consultation)
                }
            }

            binding.root.setOnClickListener {
                if (data.roomType == ChatRoomType.consultAi) {
                    val bundle =
                        bundleOf(
                            BundleKey.ROOM_ID.key() to data.serviceNumberRoomId,
                            BundleKey.SERVICE_NUMBER_ID.key() to data.serviceNumberId,
                            BundleKey.CONSULT_AI_ID.key() to data.roomId
                        )
                    IntentUtil.launchIntent(it.context, ConsultAIActivity::class.java, consultAILauncher, bundle)
                } else {
                    CoroutineScope(Dispatchers.IO).launch {
                        ChatRoomReference.getInstance().updateUnread(data.roomId)
                        withContext(Dispatchers.Main) {
                            notifyItemChanged(absoluteAdapterPosition)
                        }
                    }

                    IntentUtil.startIntent(
                        it.context,
                        ChatActivity::class.java,
                        bundleOf(BundleKey.EXTRA_SESSION_ID.key() to data.roomId)
                    )
                }
                clearUnread(data)
            }
        }
    }

    private fun clearUnread(data: SmallRoomData) {
        data.unReadNum = 0
        notifyItemChanged(currentList.indexOf(data))
    }

    @SuppressLint("SetTextI18n")
    inner class TodoSmallIconViewHolder(
        val binding: ItemAdvisoryTodoBinding
    ) : RecyclerView.ViewHolder(binding.root) {
        fun bind(data: SmallRoomData) {
            if (data.unReadNum > 0) {
                binding.tvUnread.text = data.unReadNum.toString()
                binding.tvUnread.visibility = View.VISIBLE
            } else {
                binding.tvUnread.text = ""
                binding.tvUnread.visibility = View.GONE
            }

            binding.ivTodoIcon.setImageResource(
                if (ThemeHelper.isGreenTheme() || ThemeHelper.isServiceRoomTheme) {
                    R.drawable.icon_todo_small_service_room
                } else {
                    R.drawable.icon_todo_small_normal_room
                }
            )

            binding.root.setOnClickListener {
                IntentUtil.startIntent(
                    it.context,
                    ChatRoomFilterActivity::class.java,
                    bundleOf(
                        BundleKey.INTENT_TO_TODO.key() to true,
                        BundleKey.ROOM_ID.key() to data.roomId
                    )
                )
            }
        }
    }

    class AdvisoryRoomAdapterDiffCallBack : DiffUtil.ItemCallback<SmallRoomData>() {
        override fun areItemsTheSame(
            oldItem: SmallRoomData,
            newItem: SmallRoomData
        ): Boolean = oldItem == newItem

        override fun areContentsTheSame(
            oldItem: SmallRoomData,
            newItem: SmallRoomData
        ): Boolean =
            oldItem.roomId == newItem.roomId &&
                oldItem.serviceNumberRoomId == newItem.serviceNumberRoomId &&
                oldItem.avatarId == newItem.avatarId &&
                oldItem.unReadNum == newItem.unReadNum &&
                oldItem.todoList.size == newItem.todoList.size
    }
}

data class SmallRoomData(
    val roomType: ChatRoomType,
    val roomId: String,
    var avatarId: String = "",
    var unReadNum: Int = 0,
    val serviceNumberRoomId: String = "",
    val serviceNumberId: String = "",
    val roomName: String = "",
    val todoList: List<TodoEntity> = emptyList(),
    val sort: Int = 99
)
